### ISSUE TITLE
Beam Search for RNN-T

### ISSUE THREAD

**Issue by Oktai15** [espnet/espnet#2686]
Hello!

Where can I read about Beam Search for RNN-T with more details? 

I can't understand why you generate more and more hypotheses at even on the first step of `h_i` before `len(kept_most_prob) >= beam` at here:

https://github.com/espnet/espnet/blob/97b9f888d3860717de0c44db6b48f4da9f4fe0b1/espnet/nets/beam_search_transducer.py#L163

**Comment by b-flo** [comment]
Hi, 

> Where can I read about Beam Search for RNN-T with more details?

Default beam search is equivalent to one proposed in https://arxiv.org/pdf/1211.3711.pdf, minus the prefix search part.

> I can't understand why you generate more and more hypotheses at even on the first step of h_i before len(kept_most_prob) >= beam at here:

Not sure if I understood your question correctly but if you're asking about "why we can generate more than one hypothesis at each timestep", it's because rnn-t can be expanded in an unrestricted manner, i.e. along two types of transitions (cf. previous linked paper).

**Comment by b-flo** [comment]
I'm closing this issue as the original questions are answered, feel free to ask further questions if you have any!

**Comment by Oktai15** [comment]
@b-flo ok, thanks!

I have another question. If we pretrain Prediction Network, we will do it without BLANK symbol? I suppose so, because our text data doesn't have BLANK as label. But I am confused about this moment in paper:
![image](https://user-images.githubusercontent.com/17337773/99382909-3e6b4400-28de-11eb-9a6e-fe1dcf8298a5.png)

Also there:

![image](https://user-images.githubusercontent.com/17337773/99382984-5773f500-28de-11eb-9477-e19df5d59e82.png)

The same problem with Transcription Network (Encoder).

**Comment by b-flo** [comment]
> I have another question. If we pretrain Prediction Network, we will do it without BLANK symbol? I suppose so, because our text data doesn't have BLANK as label.

No, you don't need to adapt your training data, extra symbols (blank for transducer or blank, sos and eos for mtl) are implicitly defined.

>  But I am confused about this moment in paper:

I'm not sure what's confusing you, could you be more precise? Is it because blank is also used as start/end symbols?

**Comment by Oktai15** [comment]
@b-flo, let's define vocab =  {a,b,c}, vocab_with_blank = {a, b, c, ∅}

We input in Prediction Network any sequence from vocab_with_blank or vocab? As far as I understand now, we use vocab_with_blank, because we use blank as padding/start_symbol/end_symbol for inference. It means that if we want to pretrain Prediction Network, we need to use start token, end token and padding as blank. Correct?

**Comment by b-flo** [comment]
We input in Prediction Network any sequence from vocab_with_blank or vocab? As far as I understand now, we use vocab_with_blank, because we use blank as padding/start_symbol/end_symbol for inference.

blank (or null) symbol in rnn-t is used to determine an alignment between input and output sequences during training and inference. I think you're confused about blank symbol usage and the fact it's also defined as start/end symbol in prediction network?

> It means that if we want to pretrain Prediction Network, we need to use start token, end token and padding as blank. Correct?

You would need only need the <blank> symbol!

**Comment by Oktai15** [comment]
@b-flo, here is what I figured out at this moment:

* We pretrain Prediction Network as LM using data with only one special token: blank. Blank will be as start_token and end_token in the same time for this LM during training.
* After pretraining, we use our Prediction Network in RNN-T. We will pass blank token only at once -- at the first time of inference. After that we pass only symbols from vocab.

Is it correct? Sorry for bad insight...

**Comment by b-flo** [comment]
> After pretraining, we use our RNN model as Prediction Network in RNN-T (without last linear, only RNN). We will pass blank token only at once -- at the first time of inference. After that we pass only symbols from vocab according to RNN-T inference algorithm.

I'm confused, what do you want to do? Train an initial RNN-T model and LM model and replace the prediction network by your LM for inference? Why don't you train the LM and pre-initialize prediction network weights to train an RNN-T model or use it as external LM during decoding?

During inference, prediction network will generate a symbol based on the previous one at each call. zero vector is used to start the generation

**Comment by Oktai15** [comment]
@b-flo yes, I want to understand how to train RNN-T model and how to pretrain correctly Prediction Network (no external LM at all in this discussion). As far as I know, first of all, I need to train RNN on text data. After that, I can use this RNN pretrained weights for Prediction Network and run training of full pipeline with Transcription Network, right? 

And my previous question was related to how to pre-train RNN with only blank, without EOS, BOS.

**Comment by b-flo** [comment]
> yes, I want to understand how to train RNN-T model and how to pretrain correctly Prediction Network (no external LM at all in this discussion). As far as I know, first of all, I need to train RNN on text data. After that, I can use this RNN pretrained weights for Prediction Network and run training of full pipeline with Transcription Network, right?

OK, let's assume you're using ESPnet. What you want to do is:
 - Train a RNN LM with your text data 
 - Train a RNN-T model with [transfer learning options](https://espnet.github.io/espnet/tutorial.html#how-to-use-finetuning) where you specify the LM to `--dec-init` and the modules you want to transfer to `--dec-init-mods`. The only requirement is that the prediction network architecture (layers, units) match the LM architecture for the mods you want to transfer.

**Comment by Selma-KA** [comment]
Hi,

I'm using ESPnet to train RNN-T model with and without the transfer learning option for the prediction (decoder) network but I got the same results, so I'm wondering what are the initial default values for the decoder parameters?

**Comment by b-flo** [comment]
> I'm using ESPnet to train RNN-T model with and without the transfer learning option for the prediction (decoder) network but I got the same results

Is transfer learning properly applied? We don't stop training if a wrong key is provided or if there is a mismatch, we only disable transfer learning and notify user. Training logs should show you transferred weights for given keys and warnings.

Btw, note that decoder pre-initialization with pre-trained LM is not that effective from my experiments and various reports, even less if you're using a LM trained on the same training data. Encoder pre-initialization with pre-trained CTC model is recommended though.

>  I'm wondering what are the initial default values for the decoder parameters?

If you're talking about RNNT decoder parameters initialization in ESPnet, refer to: https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transducer/initializer.py

**Comment by Selma-KA** [comment]
Thank you for the information. I noticed that in the case of mismatch, transfer learning will be disabled but I got these warnings so I think everything is fine, right?

2021-01-16 19:02:09,736 (asr_init:256) WARNING: model(s) found for pre-initialization
2021-01-16 19:02:09,736 (asr_init:200) WARNING: reading model parameters from exp/train_rnnlm_pytorch_lm/rnnlm.model.best
2021-01-16 19:02:10,961 (asr_init:240) WARNING: loading ['dec.decoder.0.weight_ih', 'dec.decoder.0.weight_hh', 'dec.decoder.0.bias_ih', 'dec.decoder.0.bias_hh', 'dec.decoder.1.weight_ih', 'dec.decoder.1.weight_hh', 'dec.decoder.1.bias_ih', 'dec.decoder.1.bias_hh'] from model: exp/train_rnnlm_pytorch_lm/rnnlm.model.best
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.0.weight_ih
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.0.weight_hh
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.0.bias_ih
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.0.bias_hh
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.1.weight_ih
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.1.weight_hh
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.1.bias_ih
2021-01-16 19:02:10,961 (asr_init:243) WARNING: override dec.decoder.1.bias_hh

**Comment by lyqlola** [comment]
@Selma-KA I think there is no problem given your log. The reason you get same results w/wo prediction network pretraining is that   the benefit it brings is quite limited, as said @b-flo.
